README.txt
// $Id: README.txt,v 1.6.2.4 2007/10/23 16:29:49 morbus Exp $

CONTENTS OF THIS FILE
---------------------

 * Introduction
 * Installation
 * IRC API Hooks
 * Design Decisions


INTRODUCTION
------------

Current Maintainer: Morbus Iff <morbus@disobey.com>

Druplicon is an IRC bot that has been servicing #drupal, #drupal-support,
and many other IRC channels since 2005, proving itself an invaluable resource.
Originally a Perl Bot::BasicBot::Pluggable application coded by Morbus Iff,
he always wanted to make the official #drupal bot an actual Drupal module.

This is the fruit of these labors. Whilst the needs of Druplicon are driving
the future and design of the module, this is intended as a generic framework
for IRC bots within Drupal, and usage outside of Druplicon is encouraged.


INSTALLATION
------------

The bot.module is not like other Drupal modules and requires a bit more
effort than normal to get going. Unlike a regular Drupal page load, an
IRC bot has to run forever and, for reasons best explained elsewhere, this
entails running the bot through a shell, NOT through web browser access.

1. This module REQUIRES Net_SmartIRC, a PHP class available from PEAR.
   In most cases, you can simply run "pear install Net_SmartIRC".

2. Copy this bot/ directory to your sites/SITENAME/modules directory.

3. Enable the module(s) and then configure them at admin/settings/bot.

4. Inside the bot/ directory is a bot_start.php script which is a wrapper
   around Drupal and the IRC network libraries. To run this script, you'll
   need to open up a shell to that directory and use the following command:

     php bot_start.php --root /path/to/drupal/root --url http://www.example.com

   --root refers to the full path to your Drupal installation directory
   and allows you to execute bot_start.php without moving it to the root
   directory. --url is required (and is equivalent to Drupal's base URL)
   to trick Drupal into thinking that it is being run as through a web
   browser. It sets HTTP_HOST and PHP_SELF, as required by Drupal.

5. Your bot is now started and is trying to connect.

6. Once your bot is connected, you can query it for more information:

     <Morbus> YOUR_BOTNAME: help

     <YOUR_BOTNAME>: Detailed information is available by asking for
       "help <feature>" where <feature> is one of: Botagotchi, Drupal URLs,
       Factoids, Seen.

     <Morbus> YOUR_BOTNAME: help Seen

     <YOUR_BOTNAME> If someone asks "seen Morbus", the bot will report the
       last time they've been seen, where, and what their last known message
       was. Directly addressing the bot will also allow the more complex
       syntax of "seen Morbus? seen d8uv?", "have you seen sbp?" and similar
       forms. * can be used as a wildcard, but only with a minimum of three
       other characters. A maximum of three results are displayed for any
       one request.

   You can also go to http://www.SITENAME.com/bot/ for a web-based version
   of all the feature help available through the IRC syntax above.


IRC API HOOKS
-------------

The following message types are supported by Net_SmartIRC:

  UNKNOWN     CHANNEL  QUERY     CTCP         NOTICE        WHO
  JOIN        INVITE   ACTION    TOPICCHANGE  NICKCHANGE    KICK
  QUIT        LOGIN    INFO      LIST         NAME          MOTD
  MODECHANGE  PART     ERROR     BANLIST      TOPIC         NONRELEVANT
  WHOIS       WHOWAS   USERMODE  CHANNELMODE  CTCP_REQUEST  CTCP_REPLY

A module may create a function in the form of:

  MODULENAME_irc_msg_MESSAGETYPE

such that a module named "bot_example" could respond or act upon all channel
messages with a function called bot_example_irc_msg_channel(). Passed to
this function is $data, an object reference that contains the message, who
said it, in what channel, and more.

Modules can respond to the user or channel with bot_message($to, $message),
where $to is either a channel name ("#drupal") or user ("Morbus Iff"). Other
IRC actions are demonstrated in the modules shipped with this package. In a
worse case scenario (ie., there's no helper function in bot.module that will
accomplish your desired tasks), you can use "global $irc;" to get the actual
Net_SmartIRC object that represents the IRC connection. Under the most ideal
conditions, you'd contribute back a patch to bot.module that'd let you
accomplish your needs without using the $irc global. Generally speaking,
try not to use the $irc global.

There is another hook available called irc_bot_reply_message (such that, in
our above example, it'd be bot_example_irc_bot_reply_message()). This function
allows you to act whenever the bot sends a message. Primarily, this was added
to allow us to log bot responses in bot_log.module. If you use this, be sure
NOT to use bot_message() within your implementation, else you'll cause an
infinite loop. Another hook, irc_bot_reply_action, does the same for actions.

The final hook of interest is irc_bot_cron, which is run every five minutes.
This is very similar to Drupal's own hook_cron, but is intended only for
bot-related operations (though, naturally, any hook_cron you add to your
own bot plugins will function as expected).

In addition to the actual utility of your module, you also should add a
few lines describing how to use your module. This is done via Drupal's
hook_help(), and the use of two special strings:

 irc:features
   Returns an array of feature names your modules provides.

 irc:features#FEATURE_NAME
   Returns an explanation of a specific feature of your module.

FEATURE_NAME will be lowercased, trimmed of whitespace, and anything not a
letter or number will be turned into an underscore. For an example in code,
take a look at the shipped bot_drupal.module. This information is provided
by the bot under the following conditions:

  <Morbus> bot_module: help

  <bot_module> Detailed information is available by asking for
    "help <feature>" where <feature> is one of:
    Drupal URLs, dns, karma.

  <Morbus> bot_module: help Drupal URLs

  <bot_module> Displays the title of drupal.org URLs ...


DESIGN DECISIONS
----------------

 * We do not enforce command prefixes such as !.

 * We do not enforce direct addressing like "botname: <command>".

Since the entire raw IRC message is passed off to each module, you are more
than welcome to enforce either of the above in your own code. Note that you
WILL have to insure that "botname: <command>" and simply "<command>" both do
as you'd expect - we do not remove "botname:" from the start of messages (and
thus a simple test of "^command$" will fail if the bot is addressed). Most
of our shipped modules cater to these two possibilities.

This is an IRC bot... in PHP. PHP is not especially awesome with regards to
memory management, and it certainly wasn't intended to run a script for any
respectable period of time (like, say, longer than the default 30 seconds).
Likewise, there's no way to uninclude a file, so any change to the loaded
modules (either codewise or enabled/disabled) will require the bot to be
restarted entirely.

Love the limitations, and craziness, of this project.


CHANGELOG.txt
// $Id: CHANGELOG.txt,v 1.14.2.16.2.43 2010/05/12 22:00:44 morbus Exp $

Bot 6.x-1.x, xxxx-xx-xx
-----------------------
  * #477596: Fixed regexp error in bot_seen.module (thanks Gurpartap).
  * bot_tell.module can now remind you of things: "BOTNAME: help Reminders?"
  * bot_potpourri.module added (thanks, ae1): "BOTNAME: help Timezones?"
  * bot_aggregator.module added (thanks cwgordon7/snufkin). Additions:
    - Configuration has been moved to the feed config form, not our own.
    - Different feeds can now be sent to different channels, based on config.
  * If we receive an error containing "flood", we slow down our sends.
  * If the server or client library receives an error, we'll watchdog it.
  * If a channel bans us, it is now permanently removed from the join list.
  * #648606: Chinese "ni hao" is mis-spelled in greetings (thanks JohnAlbin).
  * Advanced/debugging option added: whether to use real sockets or not.
  * bot_project: fixed broken Trac metadata from ticket/revision exports.
  * bot_project: function lookups now allow dashes in the branch name.
  * irc_bot_cron_faster (one minute) and _fastest (15 seconds) are available.
  * #564524: bot_project now reports comment counts on d.o URLs (thanks killes).
  * bot_tell now accepts :;, after a told nick (thanks webchick).
  * Channels are now joined every 15 seconds instead of at initial connection.
  * Logging checks for channels are now case-insensitive (thanks mozillamonks).
  * #362661: Attempt to GHOST and IDENTIFY on nick clashes (thanks seutje).
  * #380330: Added perms for factoids, karma, and project (thanks lut4rp).
  * #391916: bot_log needed another index for speedier table SELECTs.

Bot 6.x-1.1, 2009-02-10
-----------------------
  * bot_seen ping prevention matched inside strings; now only word boundaries.
  * #284666: We now use preg_quote() for various nickname escaping.
  * #349245: bot_tell doesn't consume username whitespace (thanks drewish).
  * #356003: bot_tell sorts queued messages by oldest first (thanks litwol).
  * #343245: Better regex for usernames like betz--; supports betz---- now.
  * #338723: Missing decode_entities() on project statuses (thanks RobLoach).
  * #313025: Better regex for log filtering to prevent substring matches.
  * #300206: Better factoid-ignoring of tell-like messages (thanks RobLoach).
  * #275042: Randomized messages now centralized in bot_randomized_choice().
  * #274888: Move all INSERTs and UPDATEs to use drupal_write_record().
  * #218577: bot_tell.module added (thanks Rob Loach). Additions/changes:
    - pending message queue now exists to remove SELECT on every message.
    - received messages use format_interval(), not a date (thanks litwol).
  * bot_factoid: PM a factoid with "BOTNAME: tell <who> about <factoid>".
  * #190825: Get URL to current logged discussion with "BOTNAME: log pointer?"
  * bot_name_regexp() now exists for matching inside a regular expression.
    - #117876: if bot has a nick clash, it now responds to both nicks.
    - #184015: bot name with non-word characters failed regexp addressing.
  * #137171: bot_karma.module added (thanks walkah/Rob Loach). Additions:
    - patch supported only words: committed version supports phrases.
    - if someone tries to karma themselves, the response is customizable.
    - drupal_write_record() is your friend; get used to using it!
    - "BOTNAME: karma foo?" is required to prevent bad parsing.
    - highest/lowest karmas are available at example.com/bot/karma.
    - terms less than 3 and more than 15 characters are ignored.
  * #267560: OS-specific newlines broke comparisons (thanks Gurpartap Singh).
  * #245610: bot_agotchi greeting triggers now customizable (thanks Alan Evans).
  * #184032: ignore improper hook_help implementations (thanks John Morahan).
  * #229880: bot_factoid stopwords were case sensitive (thanks John Morahan).
  * #187137: Drupal 7 style concats, and other style fixes (thanks dmitrig01).
  * #167097: fixed undefined index and better host check (thanks czarphanguye).
  * #142812: auto reconnect and retry are now configurable (Morbus/Shiny).

Bot 6.x-1.0, 2008-05-01
-----------------------
  * First release for Drupal 6.x; equivalent to 5.x-1.0.

Bot 5.x-1.0, 2008-04-08
-----------------------
  * #193209: an attempt at solving seen pings has been implemented.
  * #155045: bot_seen no longer fails with "asd%" on fuzzy matching.
  * #106993: bot_seen ignores messages and actions sent to the bot.
  * #106992: bot_seen now logs actions as seenable offenses.
  * added hook_uninstall() to everything.

Bot 5.x-1.x, 2007
-----------------
  * bot_log.module added; Logs multiple channels, day browsing, etc.
  * bot_project now has Trac integration (for tickets and revisions).
  * added another MySQL timeout preventer, thanks to killes.
  * bot_project now has configurable project.module scraping.
  * bot_project can now define URL endpoints per api branch.
  * bot_drupal renamed to bot_project, made much more generic.
  * bot: supports server passwords and channel keys, thanks to Bdragon.
  * bot_drupal: better handling of improper Doxygen summaries.
  * bot_drupal from Bdragon: integrate with api.module.
  * bot_factoid.module added; an infobot clone with all the fixins.
  * hook_irc_bot_cron() now exists and runs every five minutes.
  * #136856 and #136870 from Shiny: various PostgreSQL quote fixes.
  * bot_seen: now properly reports when it hasn't seen something.
  * bot_seen: now supports wildcard searches with * (three char minimum).
  * bot_seen: freeform "seen?" syntax is moved to direct addressing only.
  * hook_irc_bot_reply() exists to facilitate logging of own messages.
  * Issue metadata and IRC count per suggestions by ChrisKennedy.
  * bot_drupal: "Bot Drupal URLs" renamed to just "Drupal URLs".
  * bot: bot/ URL now exists and displays connection/feature details.
  * bot_agotchi: greeting and thankful responses added, user configurable.

run_bot.php
<?php
// $Id: run_bot.php,v 1.2 2006/11/11 04:39:28 morbus Exp $

/*
 * @file
 * A simple wrapper around the Drupal bootstrap and Net_SmartIRC.
 * The first half is devoted to fooling Drupal into thinking it's being run
 * through the web, and the second half gets the IRC library hooked into
 * bot.module in a Drupal-friendly hook-like way.
 */

set_time_limit(0); // ignore max_execution_time.
$script_name = array_shift($_SERVER['argv']);

// --root allows us to keep run_bot.php in the bot.module directory
// without moving it to the root of the Drupal install, and --url
// is required to trick Drupal into thinking it's a web request.
// --url should be the same thing as your Drupal $base_url.
while ($param = array_shift($_SERVER['argv'])) {
  switch ($param) {
    case '--root':
      $drupal_root = array_shift($_SERVER['argv']);
      if (is_dir($drupal_root)) { chdir($drupal_root); }
      else { die("ERROR: $drupal_root not found.\n"); }
      break;

    case '--url':
      $drupal_base_url = parse_url(array_shift($_SERVER['argv']));
      $_SERVER['HTTP_HOST'] = $drupal_base_url['host'];
      $_SERVER['PHP_SELF'] = $drupal_base_url['path'].'/'.$script_name;
      $_SERVER['REQUEST_URI'] = $_SERVER['PHP_SELF'];
      $_SERVER['REMOTE_ADDR'] = NULL; // any values here do...
      $_SERVER['REQUEST_METHOD'] = NULL; // ...odd things. uh huh.
      break;
  }
}

// load in our required libraries.
require_once './includes/bootstrap.inc';
drupal_bootstrap(DRUPAL_BOOTSTRAP_FULL);
require_once('Net/SmartIRC.php');

// initialize the bot based on sane defaults.
global $irc; $bot = new run_bot_wrapper(); $irc = new Net_SmartIRC();
$irc->setDebug( variable_get('bot_debugging', 0) ? SMARTIRC_DEBUG_ALL : SMARTIRC_DEBUG_NONE );
$irc->setAutoReconnect(TRUE);  // reconnect to the server if disconnected.
$irc->setAutoRetry(TRUE);      // retry if a server connection fails.
$irc->setChannelSyncing(TRUE); // keep a list of joined users per channel.
$irc->setUseSockets(TRUE);     // uses real sockets instead of fsock().

// send every message type the library supports to our wrapper class.
$irc->registerActionhandler(SMARTIRC_TYPE_CHANNEL, '^.*$', $bot, 'bot_class_irc_msg_channel');

// connect and begin listening.
$irc->connect( variable_get('bot_server', 'irc.freenode.net'), variable_get('bot_server_port', 6667) );
$irc->login( variable_get('bot_nickname', 'bot_module'), variable_get('bot_nickname', 'bot_module').' :http://drupal.org/project/bot', 8, variable_get('bot_nickname', 'bot_module') );
$irc->join(preg_split('/\s*,\s*/', variable_get('bot_channels', '#test')));
$irc->listen(); // go into the forever loop - no code after this is run.
$irc->disconnect(); // if we stop listening, disconnect properly.

// accept all the message type that the IRC library knows about
// and pass it off to our master bot hooks, which in turn passes
// it off to all the enabled plugin modules via Drupal's hook system.
class run_bot_wrapper {
  function bot_class_irc_msg_channel(&$irc, &$data) { bot_invoke_irc_msg_channel($data); }
}



bot_start.php
<?php
// $Id: bot_start.php,v 1.2.2.4.2.4 2009/03/05 17:41:56 morbus Exp $

/*
 * @file
 * Wrapper around the Drupal bootstrap and Net_SmartIRC. The first half is
 * devoted to fooling Drupal into thinking it's being run through the web,
 * and the other half inserts the IRC library into Drupal's hook system.
 */

set_time_limit(0); // ignore max_execution_time.
$script_name = array_shift($_SERVER['argv']);

// --root allows us to keep this script in the bot.module directory
// without moving it to the root of the Drupal install, and --url
// is required to trick Drupal into thinking it's a web request.
// --url should be the same thing as your Drupal $base_url.
while ($param = array_shift($_SERVER['argv'])) {
  switch ($param) {
    case '--root':
      $drupal_root = array_shift($_SERVER['argv']);
      is_dir($drupal_root) ? chdir($drupal_root) : exit("ERROR: $drupal_root not found.\n");
      break;

    case '--url':
      $drupal_base_url = parse_url(array_shift($_SERVER['argv']));
      if (!$drupal_base_url || !$drupal_base_url['host']) { exit("ERROR: No URL was passed via --url.\n"); }
      $drupal_base_url['path'] = isset($drupal_base_url['path']) ? $drupal_base_url['path'] : '/';
      $_SERVER['HTTP_HOST'] = $drupal_base_url['host']; // this is all very boring, no?
      $_SERVER['PHP_SELF'] = $drupal_base_url['path'] . '/' . $script_name;
      $_SERVER['REQUEST_URI'] = $_SERVER['SCRIPT_NAME'] = $_SERVER['PHP_SELF'];
      $_SERVER['REMOTE_ADDR'] = NULL; // any values here do rather...
      $_SERVER['REQUEST_METHOD'] = NULL; // ...odd things. uh huh.
      break;
  }
}

// load in our required libraries.
require_once './includes/bootstrap.inc';
drupal_bootstrap(DRUPAL_BOOTSTRAP_FULL);
require_once('Net/SmartIRC.php');

// prevent MySQL timeouts on slow channels.
db_query('SET SESSION wait_timeout = %d', 24*60*60);

// initialize the bot with some sane defaults.
global $irc; // allow it to be slurped by Drupal modules if need be.
$bot = new drupal_wrapper(); // wrapper that integrates with Drupal hooks.
$irc = new Net_SmartIRC(); // MmmmmmM. The IRC object itself. Magick happens here.
$irc->setDebug( variable_get('bot_debugging', 0) ? SMARTIRC_DEBUG_ALL : SMARTIRC_DEBUG_NONE );
// the (boolean) here is required, as Net_SmartIRC doesn't respect a FAPI checkbox value of 1, only TRUE.
$irc->setAutoReconnect((boolean)variable_get('bot_auto_reconnect', 1)); // reconnect to the server if disconnected.
$irc->setAutoRetry((boolean)variable_get('bot_auto_retry', 1)); // retry if a server connection fails.
$irc->setChannelSyncing(TRUE); // keep a list of joined users per channel.
$irc->setUseSockets(TRUE); // uses real sockets instead of fsock().

// send every message type the library supports to our wrapper class.
// we can automate the creation of these actionhandlers, but not the
// class methods below (only PHP 5 supports default methods easily).
$irc_message_types = array(
  'UNKNOWN',    'CHANNEL', 'QUERY',    'CTCP',        'NOTICE',       'WHO',
  'JOIN',       'INVITE',  'ACTION',   'TOPICCHANGE', 'NICKCHANGE',   'KICK',
  'QUIT',       'LOGIN',   'INFO',     'LIST',        'NAME',         'MOTD',
  'MODECHANGE', 'PART',    'ERROR',    'BANLIST',     'TOPIC',        'NONRELEVANT',
  'WHOIS',      'WHOWAS',  'USERMODE', 'CHANNELMODE', 'CTCP_REQUEST', 'CTCP_REPLY',
);

foreach ($irc_message_types as $irc_message_type) {
  $irc->registerActionhandler(constant('SMARTIRC_TYPE_' . $irc_message_type), '.*', $bot, 'invoke_irc_msg_' . strtolower($irc_message_type));
}

// set up a five minute timer similar to Drupal's hook_cron().
$irc->registerTimehandler(300000, $bot, 'invoke_irc_bot_cron');

// connect and begin listening.
$irc->connect(variable_get('bot_server', 'irc.freenode.net'), variable_get('bot_server_port', 6667));
$irc->login(variable_get('bot_nickname', 'bot_module'), variable_get('bot_nickname', 'bot_module') . ' :http://drupal.org/project/bot', 8, variable_get('bot_nickname', 'bot_module'), (variable_get('bot_password', '') != '') ? variable_get('bot_password', '') : NULL);

// to support passwords, we have to make a single join per channel.
$channels = preg_split('/\s*,\s*/', variable_get('bot_channels', '#test'));
foreach ($channels as $channel) {
  $channel_parts = explode(' ', $channel); // use a password if it's defined.
  $irc->join($channel_parts[0], isset($channel_parts[1]) ? $channel_parts[1] : NULL);
}

$irc->listen(); // go into the forever loop - no code after this is run.
$irc->disconnect(); // if we stop listening, disconnect properly.

// pass off IRC messages to our modules via Drupal's hook system.
class drupal_wrapper {
  function invoke_irc_bot_cron(&$irc)                 { module_invoke_all('irc_bot_cron'); }
  function invoke_irc_msg_unknown(&$irc, &$data)      { module_invoke_all('irc_msg_unknown', $data); }
  function invoke_irc_msg_channel(&$irc, &$data)      { module_invoke_all('irc_msg_channel', $data); }
  function invoke_irc_msg_query(&$irc, &$data)        { module_invoke_all('irc_msg_query', $data); }
  function invoke_irc_msg_ctcp(&$irc, &$data)         { module_invoke_all('irc_msg_ctcp', $data); }
  function invoke_irc_msg_notice(&$irc, &$data)       { module_invoke_all('irc_msg_notice', $data); }
  function invoke_irc_msg_who(&$irc, &$data)          { module_invoke_all('irc_msg_who', $data); }
  function invoke_irc_msg_join(&$irc, &$data)         { module_invoke_all('irc_msg_join', $data); }
  function invoke_irc_msg_invite(&$irc, &$data)       { module_invoke_all('irc_msg_invite', $data); }
  function invoke_irc_msg_action(&$irc, &$data)       { module_invoke_all('irc_msg_action', $data); }
  function invoke_irc_msg_topicchange(&$irc, &$data)  { module_invoke_all('irc_msg_topicchange', $data); }
  function invoke_irc_msg_nickchange(&$irc, &$data)   { module_invoke_all('irc_msg_nickchange', $data); }
  function invoke_irc_msg_kick(&$irc, &$data)         { module_invoke_all('irc_msg_kick', $data); }
  function invoke_irc_msg_quit(&$irc, &$data)         { module_invoke_all('irc_msg_quit', $data); }
  function invoke_irc_msg_login(&$irc, &$data)        { module_invoke_all('irc_msg_login', $data); }
  function invoke_irc_msg_info(&$irc, &$data)         { module_invoke_all('irc_msg_info', $data); }
  function invoke_irc_msg_list(&$irc, &$data)         { module_invoke_all('irc_msg_list', $data); }
  function invoke_irc_msg_name(&$irc, &$data)         { module_invoke_all('irc_msg_name', $data); }
  function invoke_irc_msg_motd(&$irc, &$data)         { module_invoke_all('irc_msg_motd', $data); }
  function invoke_irc_msg_modechange(&$irc, &$data)   { module_invoke_all('irc_msg_modechange', $data); }
  function invoke_irc_msg_part(&$irc, &$data)         { module_invoke_all('irc_msg_part', $data); }
  function invoke_irc_msg_error(&$irc, &$data)        { module_invoke_all('irc_msg_error', $data); }
  function invoke_irc_msg_banlist(&$irc, &$data)      { module_invoke_all('irc_msg_banlist', $data); }
  function invoke_irc_msg_topic(&$irc, &$data)        { module_invoke_all('irc_msg_topic', $data); }
  function invoke_irc_msg_nonrelevant(&$irc, &$data)  { module_invoke_all('irc_msg_nonrelevant', $data); }
  function invoke_irc_msg_whois(&$irc, &$data)        { module_invoke_all('irc_msg_whois', $data); }
  function invoke_irc_msg_whowas(&$irc, &$data)       { module_invoke_all('irc_msg_whowas', $data); }
  function invoke_irc_msg_usermode(&$irc, &$data)     { module_invoke_all('irc_msg_usermode', $data); }
  function invoke_irc_msg_channelmode(&$irc, &$data)  { module_invoke_all('irc_msg_channelmode', $data); }
  function invoke_irc_msg_ctcp_request(&$irc, &$data) { module_invoke_all('irc_msg_ctcp_request', $data); }
  function invoke_irc_msg_ctcp_reply(&$irc, &$data)   { module_invoke_all('irc_msg_ctcp_reply', $data); }
}

